File: test/rspec/producto.rb
# encoding: utf-8
require_relative 'app'
require 'json'

def listar
  RSpec.describe App do
    describe "1. Listar productos: " do
      it '1.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '1.2 Listar productos' do
        url = 'producto/listar'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
    end
  end
end

def crear
  RSpec.describe App do
    describe "2. Crear productos: " do
      it '2.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '2.2 Crear productos' do
        data = {
          :nuevos => [
            {
              :_id => 'tablaProducto_481',
              :nombre => 'Producto Corbett 1',  
              :precio_referencial => 2.13, 
              :imagen_id => nil, 
            },
            {
              :_id => 'tablaProducto_482',
              :nombre => 'Producto Corbett 2',  
              :precio_referencial => 3.20, 
              :imagen_id => nil, 
            },
            {
              :_id => 'tablaProducto_483',
              :nombre => 'Producto Corbett 3',  
              :precio_referencial => 4, 
              :imagen_id => nil, 
            },
            {
              :_id => 'tablaProducto_484',
              :nombre => 'Producto Corbett 4',  
              :precio_referencial => 0.1, 
              :imagen_id => nil, 
            },
          ],
          :editados => [],  
          :eliminados => [],
          :extra => {
          }
        }.to_json
        url = 'producto/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los productos')
        expect(test.response.body).to include('nuevo_id')
        expect(test.response.body).to include('success')
      end
    end
  end
end

def editar
  RSpec.describe App do
    describe "3. Editar productos: " do
      it '3.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '3.2 Editar productos' do
        data = {
          :nuevos => [],
          :editados =>[ 
            {
              :_id => '257511',
              :nombre => 'Producto Editado Corbett 1',  
              :precio_referencial => 2.13, 
              :imagen_id => nil, 
            },
            {
              :_id => '257501',
              :nombre => 'Producto Editado Corbett 2',  
              :precio_referencial => 3.20, 
              :imagen_id => nil, 
            },
            {
              :_id => '257508',
              :nombre => 'Producto Editado Corbett 3',  
              :precio_referencial => 4, 
              :imagen_id => nil, 
            },
            {
              :_id => '257505',
              :nombre => 'Producto Editado Corbett 4',  
              :precio_referencial => 0.1, 
              :imagen_id => nil, 
            }
          ],  
          :eliminados => [],
          :extra => {
          }
        }.to_json
        url = 'producto/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los productos')
        expect(test.response.body).to include('success')
      end
    end
  end
end

def eliminar
  RSpec.describe App do
    describe "4. Eliminar productos: " do
      it '4.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '4.2 Eliminar productos' do
        data = {
          :nuevos => [],
          :editados =>[],  
          :eliminados => ['257511','257501','257508','257505'],
          :extra => {
          }
        }.to_json
        url = 'producto/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los productos')
        expect(test.response.body).to include('success')
      end
    end
  end
end


#listar
#crear
#editar
eliminar

File: views/producto.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
from bottle import Bottle, request
from pprint import pprint
from datetime import datetime
from config.database import db

producto_view = Bottle()

@producto_view.route('/listar', method='GET')  
def listar():
  rpta = []
  locales = db.collection('productos')
  for local in locales:
    try:
      rpta.append({
        '_id': local['_key'], 
        'nombre': local['nombre'], 
        'precio_referencial': local['precio_referencial'], 
        'imagen_id': local['imagen_id'], 
      })
    except KeyError as e:
      print('KeyError listar productos en ArangoDBJSON, la llave que busca ' + str(e) + ' no existe')
  return json.dumps(rpta)

@producto_view.route('/guardar', method='POST')
def guardar():
  data = json.loads(request.query.data)
  nuevos = data['nuevos']
  editados = data['editados']
  eliminados = data['eliminados']
  array_nuevos = []
  rpta = None
  try:
    if len(nuevos) != 0:
      for nuevo in nuevos:
        _id = db.collection('productos').insert({
          'nombre': nuevo['nombre'], 
          'precio_referencial': nuevo['precio_referencial'], 
          'imagen_id': nuevo['imagen_id'], 
          'modifiacion': None, 
          'creacion': datetime.now().__str__()
        })
        temp_id = nuevo['_id']
        temp = {'temporal': temp_id, 'nuevo_id': _id['_key']}
        array_nuevos.append(temp)
    txn = db.transaction(write = 'productos')
    if len(editados) != 0:
      for editado in editados:
        txn.collection('productos').update({
          '_key': editado['_id'], 
          'nombre': editado['nombre'], 
          'precio_referencial': editado['precio_referencial'], 
          'imagen_id': editado['imagen_id'], 
          'modifiacion': datetime.now().__str__(), 
        })
    if len(eliminados) != 0:
      for _key in eliminados:
        txn.collection('productos').delete({
          '_key': _key,  
        })
    rpta = {'tipo_mensaje' : 'success', 'mensaje' : ['Se ha registrado los cambios en los productos', array_nuevos]}
    txn.commit()
  except Exception as e:
    if len(array_nuevos) != 0:
      txn = db.transaction(write = 'productos')
      for temp in array_nuevos:
        _key = temp['nuevo_id']
        txn.collection('productos').delete({
          '_key': _key,  
        })
      txn.commit()
    rpta = {'tipo_mensaje' : 'error', 'mensaje' : ['Se ha producido un error en guardar los productos', str(e)]}
  return json.dumps(rpta)